cmd/data-node/commands/start/node_pre.go

```diff
@@ -233,21 +233,28 @@ func (l *NodeCommand) preRun([]string) (err error) {
 
 	eventSource = broker.NewFanOutEventSource(eventSource, l.conf.SQLStore.FanOutBufferSize, 2)
 
-	var onBlockCommittedFn func(ctx context.Context, chainId string, lastCommittedBlockHeight int64)
+	var onBlockCommittedHandler func(ctx context.Context, chainId string, lastCommittedBlockHeight int64)
+	var protocolUpgradeHandler broker.ProtocolUpgradeHandler
 
-	if l.snapshotService != nil {
+	if l.conf.DeHistory.Enabled {
 		blockCommitHandler := snapshot.NewBlockCommitHandler(l.Log, l.snapshotService.SnapshotData, l.networkParameterService.GetByKey,
-			l.conf.Broker)
-		onBlockCommittedFn = blockCommitHandler.OnBlockCommitted
+			bool(l.conf.Broker.UseEventFile), l.conf.Broker.FileEventSourceConfig.TimeBetweenBlocks.Duration)
+		onBlockCommittedHandler = blockCommitHandler.OnBlockCommitted
+		protocolUpgradeHandler = dehistory.NewProtocolUpgradeHandler(l.Log, l.conf.DeHistory, l.protocolUpgradeService,
+			l.snapshotService.CreateSnapshotAsynchronously, l.deHistoryService.SnapshotPublished)
 	} else {
-		onBlockCommittedFn = func(ctx context.Context, chainId string, lastCommittedBlockHeight int64) {}
+		onBlockCommittedHandler = func(ctx context.Context, chainId string, lastCommittedBlockHeight int64) {}
+		protocolUpgradeHandler = dehistory.NewProtocolUpgradeHandler(l.Log, l.conf.DeHistory, l.protocolUpgradeService,
```

Comment:
Why can't `NewProtocolUpgradeHandler()` just take in `l.snapshotService` and `l.dehistoryService` to call those functions? Its all a bit convoluted passing their functions around. The config is save in it already so you just need to do `l.conf.DeHistory.Enabled` as a condition on whether to call them:
```
func (t *ProtocolUpgradeHandler) OnProtocolUpgradeEvent() {
    if !t.cfg.Enabled {
        t.protocolUpgradeService.SetProtocolUpgradeStarted()
        return
   }

    // do the snapshot stuff
    t.snapshotService.createSnapshotAsynchronously()
}

Reply:
Protocol handler has changed, this comment now out of date